_posts/2016-08-09-the-tech-behind-live-auction-integration.markdown

```diff
@@ -0,0 +1,57 @@
+
+Causality was developed in Scala, using the [Akka](http://doc.akka.io/docs/akka/current/intro/what-is-akka.html) technology suite for distributed computing. At its core is an append-only storage engine, based on [Akka Persistence](http://doc.akka.io/docs/akka/current/scala/persistence.html), with a small library we developed called [Atomic Store](https://github.com/artsy/atomic-store) that allowed us to achieve strict consistency, at the cost of maximal throughput. This is a trade-off that is explored in the readme of that project, and is sensible for our application. A layer of auction business logic validates prospective changes in state (e.g. new bids, edits to parameters) and produces the derived view of the append-only event log.
+
+Lastly, Causality has an [Akka HTTP](http://doc.akka.io/docs/akka/current/scala/http/introduction.html)-based API layer, with a WebSocket server implemented using [Akka Streams](http://doc.akka.io/docs/akka/current/scala/stream/stream-introduction.html). Asynchronous updates generated in the event processing logic are published across the cluster using [Akka Distributed Pub/Sub](http://doc.akka.io/docs/akka/current/scala/distributed-pub-sub.html), and they are merged into the WebSocket outflow.
+
+# Pre-bidding, tooling, and other concerns
+
+In addition to accepting bids during a live sale, we also allow users to place bids before the event begins. In practice, this is almost the same workflow as our existing timed auction experience. For this reason, we chose to leverage all of our existing technology, including that which was used for state management. The work of preparing our preexisting tech for LAI involved a number of cosmetic changes to our front-end UI and messaging to make them appropriate for a live sale, as well as a reliable handoff of responsibility from these preexisting front- and back-end services to the new ones at the time the sale goes live. Additionally, our tooling for setting up, administering, monitoring, and closing out sales had to be adapted to account for the differences of LAI sales. We made tactical decisions of where to update our tooling and where to manually work around these mismatches.
+
+We relied on our automated test suites, as well as thorough manual testing, to ensure that this handoff functioned smoothly under various circumstances. We  will eventually eliminate this duplication. But this will require delicate refactoring of our preexisting tooling, which we will take on, even as we execute a fall auction season significantly busier than the last.
```

Comment:
The manual testing was really key, thanks for pointing it out. Could be worth a shout-out to our auctions ops team, especially Katarina and Devang, who really helped us with all the testing.